syne_tune/optimizer/schedulers/searchers/bayesopt/models/sklearn_predictor.py

```diff
@@ -27,36 +27,31 @@
 
 class SKLearnPredictorWrapper(BasePredictor):
     """
-    Wrapper class for the sklearn estimators to be used with ContributedEstimatorWrapper
+    Wrapper class for sklearn predictors returned by ``fit_from_state`` of
+    :class:`~syne_tune.optimizer.schedulers.searchers.bayesopt.models.sklearn_estimator.SKLearnEstimatorWrapper`.
     """
 
     def __init__(
         self,
-        contributed_predictor: SKLearnPredictor,
+        sklearn_predictor: SKLearnPredictor,
         state: TuningJobState,
         active_metric: Optional[str] = None,
     ):
         super().__init__(state, active_metric)
-        self.contributed_predictor = contributed_predictor
+        self.sklearn_predictor = sklearn_predictor
 
-    def predict(self, inputs: np.ndarray) -> List[Dict[str, np.ndarray]]:
-        """
-        Returns signals which are statistics of the predictive distribution at
-        input points ``inputs``. By default:
-
-        * "mean": Predictive means.
-        * "std": Predictive stddevs, shape ``(n,)``
-
-        This function relies on the assigned ContributedPredictor to execute the predictions
+    def keys_predict(self) -> Set[str]:
+        if self.sklearn_predictor.returns_std():
+            return {"mean", "std"}
+        else:
+            return {"mean"}
 
-        :param inputs: Input points, shape ``(n, d)``
-        :return: List of ``dict`` with keys :meth:`keys_predict`, of length 1
-        """
-
-        mean, std = self.contributed_predictor.predict(inputs)
-        outputs = {"mean": mean}
-        if std is not None:
-            outputs["std"] = std
+    def predict(self, inputs: np.ndarray) -> List[Dict[str, np.ndarray]]:
+        prediction = self.sklearn_predictor.predict(X=inputs)
+        if self.sklearn_predictor.returns_std():
+            outputs = {"mean": prediction[0], "std": prediction[1]}
+        else:
+            outputs = {"mean": prediction}
```

Comment:
We can do that, but then I think you have not really simplified things for the user.

My idea is that code in `bayesopt/sklearn` is simpler and closer to scikit-learn that code in `bayesopt/models`. Here, "simpler" means that it looks more like scikit-learn, and less like the specific API we need further above.

You started this way, and I like it (except the fact whether a predictor returns std or not, cannot be an argument, but a fixed property of the predictor).

If another case comes up, I'd rather create a different class for predictor and estimator in `bayesopt/sklearn`.

Reply:
Up to you, but just pushing through all methods to `bayesopt/sklearn` does not make things easier there.

Reply:
I think this is a good point but then we should commit to making it as simple as possible and only accept mean + std, like in the original proposal. Adding options to only sometimes return std makes it more contrived and defeats the purpose of making is as simple as possible.

In that spirit, we would add different wrappers/predictors for quantile output as well as mean-only output. Ultimately, it is very hard to add do BO with just mean predictions so it woulnt just work out-of-the box with the simple surrogate

Reply:
Also, we already have a use case for deterministic models (namely, cost functions).

Reply:
Note that this is not an option, but a fixed property of the object (or even of the class). A predictor either always returns (mean, std), or it always only returns mean.

Reply:
I can see your point for improved flexibility but I think that with the Scikit interface the basic class should just be as simple to use as possible. This means just implementing `fit()` and `predict()` methods should be enough. The current implementation supports providing `std=None` so non-std functions are still supported if needed, the interface is just designed to make the default use-case as easy as possible and I think it is something we should retain.

Note that we can also add a more advanced SKLearnPredictor that has higher flexibility at the cost of complexity, it just won't be the default that we showcase as easy to use.

Reply:
Well, so how do we proceed?

The current implementation is simply not correct. It is essential for things to work that `keys_predict` and the return of `predict` are consistent. As it stands, `keys_predict` says: I'll give you mean and std, but then maybe `predict` does not. An AF like EI thinks everything is fine, and the first call, things blow up.

I am all for simplicity, but things just have to be correct.

If you want `keys_predict` to remain as it is, then `predict` has to always return mean and std, and so a deterministic model cannot be covered.

I don't really understand why not being explicit about whether you return std or not, is making things complicated. This seems pretty basic to me, that you can make a statement about that.

Reply:
> The current implementation is simply not correct. It is essential for things to work that keys_predict and the return of predict are consistent. As it stands, keys_predict says: I'll give you mean and std, but then maybe predict does not. An AF like EI thinks everything is fine, and the first call, things blow up.

Aha I see, I missed this point before but this is indeed a good callout. I think in this case I would opt for the simplest way to fix the bug and ublock the tutorial: assert both mean and std are returned by the default surrogate. This is the surrogate you can use for the tutorials and it will be simple to understand albeit non-flexible.

Going beyond that: we generally need a unified way to tackle predictors that return mean, mean+std, quantiles or anything else. I see the main issue is that the user implementing their own `SKLearnPredictor` needs to somehow encode the information about the returns of the predictor inside the class. One way is to have `SKLearnPredictor` return a dictionary thus moving all the logic about different outputs into `SKLearnPredictor` (that's my previous proposal). The second one is to assume multiple options for `SKLearnPredictor` and move the construction of output dictionaries into `SKLearnPredictorWrapper` such as below which I hope will be simple in the default case but also flexible for advanced users.

```

class SKLearnPredictor:
    """
    Base class for the sklearn predictors
    """

    def predict(
        self, X: np.ndarray, return_std: bool = True
    ) -> Tuple[np.ndarray, np.ndarray]:
        """
        Returns signals which are statistics of the predictive distribution at
        input points ``inputs``.


        :param inputs: Input points, shape ``(n, d)``
        :return: Tuple with the following entries:
            * "mean": Predictive means in shape of ``(n,)``
            * "std": Predictive stddevs, shape ``(n,)``
        """

        raise NotImplementedError()


class SKLearnDeterminicticPredictor:
    """
    Base class for the sklearn predictors
    """

    def predict(
        self, X: np.ndarray, return_std: bool = True
    ) -> np.ndarray:
        """
        Returns signals which are statistics of the predictive distribution at
        input points ``inputs``.


        :param inputs: Input points, shape ``(n, d)``
        :return: "mean": Predictive means in shape of ``(n,)``
        """

        raise NotImplementedError()
    
    
class SKLearnPredictorWrapper(BasePredictor):
    def __init__(
        self,
        contributed_predictor: Union[SKLearnPredictor, SKLearnDeterministicPredictor],
        state: TuningJobState,
        active_metric: Optional[str] = None,
    ):
        super().__init__(state, active_metric)
        self.contributed_predictor = contributed_predictor
        if isinstance(self.contributed_predictor, SKLearnPredictor):
            self.keys = {"mean", "std"}
        elif isinstance(self.contributed_predictor, SKLearnDeterminicticPredictor):
            self.keys = {"mean"}
        else:
            raise AttributeError

    def predict(self, inputs: np.ndarray) -> List[Dict[str, np.ndarray]]:
        predictions = self.contributed_predictor.predict(inputs)
        if isinstance(self.contributed_predictor, SKLearnPredictor):
            assert isinstance(predictions, tuple), "SKLearnPredictor must return a tuple with Mean and Std"
            assert len(predictions) == 2, "SKLearnPredictor must return a tuple with Mean and Std"
            outputs = {"mean": predictions[0], "std": predictions[1]}
        elif isinstance(self.contributed_predictor, SKLearnDeterminicticPredictor):
            assert isinstance(predictions, np.ndarray), "SKLearnPredictor must return a  single array with the mean"
            outputs = {"mean": predictions}
        else:
            raise AttributeError

        return [outputs]
    
    def keys_predict(self) -> Set[str]:
        return self.keys
```

Reply:
Your quantile case will be interesting indeed.